src/main/java/com/google/edith/servlets/UserInsights.java

```diff
@@ -0,0 +1,217 @@
+                                (long) item.getProperty("quantity"),
+                                                               (String) item.getProperty("date")
+                              );
+                            } catch (EntityNotFoundException e) {
+                              return null;
+                            }
+                            })
+                            .collect(Collectors.toList());
+    String itemsJson = GSON.toJson(items);
+    JsonObject userJson = new JsonObject();
+    userJson.addProperty("weeklyAggregate", aggregateJson);
+    userJson.addProperty("items", itemsJson);
+    return GSON.toJson(userJson);
+  }
+
+  /**
+   * Finds the UserStats entity corresponding to {@code userId} in datastore.
+   * @return UserStats entity for this user
+   */
+  private Optional<Entity> retreiveUserStats() {
+    Filter idFilter = new FilterPredicate("userId", 
+                                FilterOperator.EQUAL,
+                                userId);
+    Query query = new Query("UserStats").setFilter(idFilter);
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> entities = results.asList(FetchOptions.Builder.withLimit(10));
+    return entities.isEmpty() ? Optional.empty() 
+                              : Optional.of(entities.get(0));
+  }
+
+  /**
+   * Creates a map relating string weekly period keys to string spending values.
+   * This method calculates the trailing based on a date. 
+   * Example: {"06-20-2020" : "53.0"}
+   * @param items - A list of {@code Key} objects that reference Item entities
+                  - in the datastore.
+   * @return Creates a map with keys for each ending day of a weekly period and  
+   *         values for the total spending during that period. 
+   * TODO (malachibre) : Modify this method by using an enum to calculate time
+   *                     period totals using an enum.
+   */
+  private ImmutableMap<String, String> calculateWeeklyTotal(List<Key> itemKeys)  {
+    if (itemKeys == null || itemKeys.isEmpty()) {
+        return ImmutableMap.copyOf(new HashMap<String, String>());
+    }
+    Map<String, String> weeklyTotals = new HashMap<String, String>();
+    LocalDate currentEndOfWeek;
+
+    try {
+      currentEndOfWeek = getEndOfWeek(LocalDate.parse((String) datastore.get(itemKeys.get(0))
+                                          .getProperty("date"), DATE_FORMATTER));
+    } catch(EntityNotFoundException e) {
+      System.err.println("Error: Entity could not be located");
+      return ImmutableMap.copyOf(new HashMap<String, String>());
+    }
+
+    double weeklyTotal = 0;
+    Entity itemEntity;
+    for(Key itemKey : itemKeys) {
+      try {
+        Entity item = datastore.get(itemKey);
+        LocalDate itemDate = LocalDate.parse((String) item.getProperty("date"),
+                                             DATE_FORMATTER);
+        if (ChronoUnit.DAYS.between(itemDate, currentEndOfWeek) < 0) {
```

Comment:
Can't find the previous comment, but is there anyway to make this condition more intuitive?

Reply:
Yes!